Avoid differing results in `linalg.(tensor_)solve`

Flamefire opened the pull request:
Remove an optimization potentially using a transposed matrix as input for `linalg_lu_factor_ex_out`.

Depending on whether the input memory layout is contiguous or not this may lead to slightly different results which may cause larger differences in subsequent steps ultimately leading to test failures in e.g. `test_vmapvjp_linalg_tensorsolve_cpu_float32` & `test_vmapvjpvjp_linalg_tensorsolve_cpu_float32`.

The intended optimization no longer applies after 59bc76f so this code can be removed too resolving the accuracy issues observed in those tests.

With this change the code path used for the "regular" and "vmap" cases are identical: A batched tensor is iterated over in the batch dimension  in [lu_solve](https://github.com/pytorch/pytorch/blob/f84062f78d723543e4962ffa1d38dcf42947e3f1/aten/src/ATen/native/BatchLinearAlgebraKernel.cpp#L1009) and [lu_factor](https://github.com/pytorch/pytorch/blob/f84062f78d723543e4962ffa1d38dcf42947e3f1/aten/src/ATen/native/BatchLinearAlgebraKernel.cpp#L955)    
Prior to this it might not be the case as either tensor would/could have been non-contiguous leading to using a transposed tensor for the LU factorization instead.

So the (CPU) results should now be identical.

Fixes #151440
Fixes #114868

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

Flamefire committed:
Avoid differing results in `linalg.(tensor_)solve`

Remove an optimization potentially using a transposed matrix as input
for `linalg_lu_factor_ex_out`.

Depending on whether the input memory layout is contiguous or not this
may lead to slightly different results which may cause larger
differences in subsequent steps ultimately leading to test failures in
e.g. `test_vmapvjp_linalg_tensorsolve_cpu_float32` & `test_vmapvjpvjp_linalg_tensorsolve_cpu_float32`.

The intended optimization no longer applies after 59bc76f so this code
can be removed too resolving the accuracy issues observed in those tests.

Fixes #151440

Flamefire commented:
@pytorchbot label "release notes: nn"

Flamefire committed:
Adjust derivative calculation for removed linalg.solve optimization

lezcano reviewed (commented):
please remove the updated third party and add a regression test.

Flamefire commented:
> please remove the updated third party and add a regression test.

Updated the commit. There are already 2 tests that fail in my environment without this and pass with: test_vmapvjpvjp_linalg_tensorsolve_cpu_float32 & test_vmapvjp_linalg_tensorsolve_cpu_float32 See https://github.com/pytorch/pytorch/issues/114868

I double-checked and although I think I've seen that failure on Intel CPUs before (at least in PyTorch 2.1.0) but it seems to succeed now (on Intel and 2.6). Some debugging into `PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=2 python functorch/test_ops.py TestOperatorsCPU.test_vmapvjp_linalg_tensorsolve_cpu_float32`:

AMD:
```
Greatest absolute difference: 0.06817626953125 at index (0, 0, 3, 0, 2)
Greatest relative difference: 0.00010866165393963456 at index (0, 1, 0, 2, 0)
```
Intel:
```
Greatest absolute difference: 0.0118408203125 at index (0, 0, 3, 1, 1)
Greatest relative difference: 1.1191742487426382e-05 at index (0, 0, 1, 2, 0)
```

And because "up to 0.0001 allowed" for both atol and rtol the latter succeeds.

As a reproducer I could add a tolerance override for test_vmapvjp:
```
            tol1(
                "linalg.tensorsolve",
                {torch.float32: tol(atol=1e-4, rtol=1e-05)},
            ),
```

This causes a failure also on Intel which is resolved after the patch.   
I wasn't able to find a tolerance that triggers the failure on Intel for test_vmapvjpvjp_linalg_tensorsolve_cpu_float32 though.

Flamefire commented:
@lezcano I tightened the tolerances as described and the PR #152180 fails as expected. I rebased this PR on top of it so the commit status is kept. Those failing tests succeed here.

Does that suffice as a regression test?

lezcano commented on test/functorch/test_ops.py:
```diff
@@ -1195,6 +1195,8 @@ def vjp_of_vjp(*args_and_cotangents):
                 "linalg.householder_product",
                 {torch.float32: tol(atol=3e-04, rtol=9e-04)},
             ),
+            # Tighten tolerance to check for regressions of issue #114868
+            tol1("linalg.tensorsolve", {torch.float32: tol(atol=1e-4, rtol=1e-5)}),
```
Yep, let's remove this one as any change in hardware may make it fail, but yeah, thank you for checking that this indeed solves the regression.

lezcano replied:
If I understand correctly, the iterative version can use a different algorithm than the batched version with very different numerical properties. If that's the case, the point above still stands.

If the two are actually calling exactly the same ops and we can guarantee bitwise equivalence, then let's set atol = rtol = 0.

Flamefire replied:
> the iterative version can use a different algorithm

This change removes one place where this was the case yes. From what I remember during my investigation of this issue this now does not use a different algorithm in any case.   
This could change in the future of course but I don't see how you'd implement linalg solver over multiple batches differently than iterating them.

> then let's set atol = rtol = 0.

We can formally do that but it has no effect: There is a minimum tolerance set depending on the data type. So the check wouldn't be for bitwise equivalence but I guess it shows the intention well enough so we could do that.

lezcano commented:
@pytorchbot merge

Flamefire commented:
Looks like a failure in trunk/main?

larryliu0820 committed:
[BE] Cleanup old ExecuTorch codegen and runtime code (#154165)

Summary: These files are added to pytorch/pytorch before ExecuTorch is
opensourced. Now is a good time to remove it from pytorch/pytorch, since
the code is moved to pytorch/executorch already.

Test Plan: Rely on CI jobs.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/154165
Approved by: https://github.com/kimishpatel, https://github.com/Skylion007, https://github.com/cyyever

ebetica committed:
Update README.md - James has the wrong github link. (#151473)

Unless I'm wrong, the James on the pytorch paper is not the account linked to in the README.md.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/151473
Approved by: https://github.com/albanD

nWEIdia committed:
[CI][CUDA][UCC] Update test_c10d_ucc.py - remove xfailIfLinux because it now succeeds (#150979)

 pytest -v test/distributed/test_c10d_ucc.py  -k test_save_load
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.12.3, pytest-8.1.1, pluggy-1.5.0 -- /usr/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/opt/pytorch/pytorch/.hypothesis/examples'))
rootdir: /opt/pytorch/pytorch
configfile: pytest.ini
plugins: anyio-4.9.0, hypothesis-6.130.13, flakefinder-1.1.0, rerunfailures-15.0, xdist-3.6.1, xdoctest-1.0.2, typeguard-4.3.0
collected 63 items / 62 deselected / 1 selected
Running 1 items in this shard

test/distributed/test_c10d_ucc.py::DistributedDataParallelTest::test_save_load_checkpoint PASSED [65.2581s]                                                                                               [100%]

================================================================================== 1 passed, 62 deselected in 68.78s (0:01:08)

@ptrblck @eqy @tinglvv @atalman @malfet

Pull Request resolved: https://github.com/pytorch/pytorch/pull/150979
Approved by: https://github.com/eqy

pytorchupdatebot committed:
[audio hash update] update the pinned audio hash (#154776)

This PR is auto-generated nightly by [this action](https://github.com/pytorch/pytorch/blob/main/.github/workflows/nightly.yml).
Update the pinned audio hash.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/154776
Approved by: https://github.com/pytorchbot

lezcano commented:
@pytorchbot merge -r

lezcano commented:
the rebase onto `viable/strict` messed up the PR. Mind resending it rebased on top of `viable/strict`?

Flamefire commented:
--> https://github.com/pytorch/pytorch/pull/154983
